# Тема 1 "Введение в математических анализ"

### 1. Как относятся друг к другу множество и последовательность? (в ответе использовать слова типа: часть, целое, общее, частное, родитель, дочерний субъект и т.д.)

Последовательность по сути своей является результат последовательного частичного выбора элементов множества, элементы которого пронумерованы натуральными числами.

### 2. Прочитать высказывания математической логики, построить их отрицания и установить истинность.

#### 2.1 $\forall y \in [0;1] : sgn(y) = 1$

*Высказывание:*

Для любого $ y $ принадлежащего сегменту $ [0;1] $ выполняется равенство $ sgn(y) = 1 $

*Истинность:*

Данное утверждение ложно, т.к. сигнум нуля равен нулю.

*Отрицание:*

$\exists y \in [0;1] : sgn(y) \neq 1$.

#### 2.2 $\forall n \in \mathbb N > 2 : \exists x, y, z \in \mathbb N : x^n = y^n + z^n$

*Высказывание:*

Для любого $ n $ из множества натуральных чисел больше двух существуют такие натуральные числа $ x, y, z $ в которых $ n $ выражается через степень полинома - икс в степени $ x^n = y^n + z^n $.

*Истинность:*

Утверждение не верно согласно *Великой теореме Ферма*, которая утверждает, что такое уравнение при `n` больше двух не имеет целых корней.

*Отрицание:*

$\exists n \in \mathbb N > 2 : \forall x, y, z \in \mathbb N : x^n \neq y^n + z^n$


#### 2.3 $\forall x \in \mathbb R \exists X \in \mathbb R : X > x$

*Высказывание:*

Для любого вещественного $ x $ существует множество вещественных чисел $ X $ больше чем $ x $.

*Истинность:*

Утверждение истинно так множество вещественных чисел бесконечно.

*Отрицание:*

$\exists x \in \mathbb R \ \forall \ X \in \mathbb R : X < x$

#### 2.4 $\forall x \in \mathbb C \nexists y \in \mathbb C : x > y || x < y$

*Высказывание:*

Для любого комплексного числа $ x $ не существует комплексного числа $ y $, которое меньше или больше $ x $.

*Истинность:*

Утверждение истинно т.к. для $ x $ = $ y $ условие ($x > y || x < y$) не выполниться.

*Отрицание:*

$\exists x \in \mathbb C \ \exists \ y \in \mathbb C : x \leqslant y || x \geqslant y$

#### 2.5 $\forall y \in [0; \dfrac{\pi}{2}] \ \exists \ \epsilon > 0 : sin y < sin(y + \epsilon) $

*Высказывание:*

Для любого $ y $ из интервала от нуля включительно до  $\dfrac{\pi}{2}$ включительно существует эпсилон больший нуля, такой, что синус $ y $ меньше чем синус  от $ y $ + $\epsilon$.

*Истинность:*

Утверждение ложно т.к. при $y = \dfrac{\pi}{2}$ и $\epsilon > 0$ синусоида $sin(y + \epsilon)$ пойдет на снижение и условие не выполнится. 

*Отрицание:*

$\exists y \in [0; \dfrac{\pi}{2}] \ \forall \ \epsilon > 0 : sin y \geqslant sin(y + \epsilon) $

#### 2.6 $\forall y \in [0; pi) \ \exists \ \epsilon > 0 : cos y < cos(y + \epsilon) $

*Высказывание:*

Для любого $ y $ из интервала от нуля включительно до $\pi$ не включительно существует эпсилон больший нуля, такой, что косинус $ y $ меньше, чем косинус от $ y $ + $\epsilon$.

*Истинность:*

Утверждение истинно так как $\pi$ не включено в интервал и условие выполнится для любого `y`.

*Отрицание:*

$\exists y \in [0; \pi) \ \forall \ \epsilon > 0 : cos y \leqslant cos(y + \epsilon) $

#### 2.7 $\in x : x \notin \{ \mathbb N, \mathbb Z, \mathbb Q, \mathbb R, \mathbb C \} $

*Высказывание:*

Существует такое $ x $ которое не включено ни в одно из множеств натуральных, целых, рациональных, вещественных и комплексных чисел.

*Истинность:*

Данное утверждение ложно так как согласно парадоксу Рассела нельзя составить множество всех множеств, а значит само условие высказывания не верно.

*Отрицание:*

$\forall x : x \in \{ \mathbb N, \mathbb Z, \mathbb Q, \mathbb R, \mathbb C \} $

# Тема 2 "Множество"


#### 1. Даны три множества a,b и с. Необходимо выполнить все изученные виды бинарных операций над всеми комбинациями множеств.



Определим множества: 

* $A = \{ 1; 4; 6\}$
* $B = \{ 2; 4; 5\}$
* $C = \{ 2; 3; 4\}$

1. Пересечение: 

$ A \cap B \cap C = \{4\} $

2. Объединение: 

$ A \cup B \cup C = \{ 1; 2; 3; 4; 5 \} $

3. Разность: 

$ A \setminus B \setminus C = \{ 1; 6 \} $

4. Симметричной разность: 

$ A \: \Delta \: B \: \Delta \: C = \{ 1; 3; 5; 6 \} $

5. Декартово произведение: 

$ A \times B \times C = \{ (1; 2; 2), (4; 4; 3) ... \} $

#### 2. *Выполнить задание 1 на языке Python

In [2]:
A = {1, 4, 6}
B = {2, 4, 5}
C = {2, 3, 4}

In [6]:
# Пересечение
A & B & C

{4}

In [7]:
# Объединение
A | B | C

{1, 2, 3, 4, 5, 6}

In [8]:
# Разность
A - B - C

{1, 6}

In [9]:
# Симметричной разность
(A - B - C) | (B - C - A) | (C - A - B)

{1, 3, 5, 6}

In [11]:
# Декартово произведение
import itertools
print(*itertools.product(A, B, C))

(1, 2, 2) (1, 2, 3) (1, 2, 4) (1, 4, 2) (1, 4, 3) (1, 4, 4) (1, 5, 2) (1, 5, 3) (1, 5, 4) (4, 2, 2) (4, 2, 3) (4, 2, 4) (4, 4, 2) (4, 4, 3) (4, 4, 4) (4, 5, 2) (4, 5, 3) (4, 5, 4) (6, 2, 2) (6, 2, 3) (6, 2, 4) (6, 4, 2) (6, 4, 3) (6, 4, 4) (6, 5, 2) (6, 5, 3) (6, 5, 4)


# Тема 3 "Последовательность"


### 1. Даны 4 последовательности.

1.1 $ \{a_n\}_{n=1}^{\infty} = 2^n - n $

1.2 $ \{b_n\}_{n=2}^{\infty} = \dfrac{1}{1 - n} $

1.3 $ \{c_n\}_{n=1}^{\infty} = -1^n + \sqrt{2n}$

1.4 $ \{d_n\}_{n=1}^{\infty} = (-1)^{2n} + \dfrac{1}{n^2}$

Необходимо:

1. исследовать их на монотонность; 
2. исследовать на ограниченность;
3. найти пятый по счету член.

#### 1.1 Решение

$$ \{a_n\}_{n=1}^{\infty} = 2^n - n $$

Найдем предел:

$$ \lim \limits_{n \to \infty}  (2^n - n) = 2^{\infty} - \infty = \infty $$

Проверим подстановкой:


In [14]:
seq1 = []
for n in range(1, 10):
    seq1.append(2**n - n)
seq1

[1, 2, 5, 12, 27, 58, 121, 248, 503]

* Последовательность монотонно возрастающая. 
* Снизу ограничена единицей, сверху не ограничена, т.е. предел положительно бесконечный.
* Пятый по счету член = 27.

#### 1.2 Решение

$$ \{b_n\}_{n=2}^{\infty} = \dfrac{1}{1 - n} $$

Найдем предел:

$$\lim \limits_{n \to \infty}  (\dfrac{1}{1 - n}) = \dfrac{1}{1 - \infty} = 0 $$

Проверим подстановкой:

In [17]:
seq2 = []
for n in range(2, 8):
    seq2.append(1/(1 - n))
seq2

[-1.0, -0.5, -0.3333333333333333, -0.25, -0.2, -0.16666666666666666]

* Последовательность монотонно возрастающая. 
* Снизу ограничена минус единицей, а сверху нулём.
* Пятый по счету член = -0.2.

#### 1.3 Решение

$$ \{c_n\}_{n=1}^{\infty} = -1^n + \sqrt{2n}$$

Найдем предел:

$$\lim \limits_{n \to \infty}  (-1^n + \sqrt{2n}) = -1^{\infty} + \sqrt{2 \cdot \infty} = -1 + \infty = \infty$$

Проверим подстановкой:

In [19]:
seq3 = []
for n in range(1, 10**5):
    seq3.append(-1**n + (2 * n)**0.5)
seq3[0:5]

[0.41421356237309515,
 1.0,
 1.4494897427831779,
 1.8284271247461903,
 2.1622776601683795]

* Последовательность монотонно возрастающая. 
* Снизу ограничена значением примерно 0.414, сверху не ограничена.
* Пятый по счету член $\approx 2.16$

#### 1.4 Решение

$$ \{d_n\}_{n=1}^{\infty} = (-1)^{2n} + \dfrac{1}{n^2}$$

Найдем предел:

$$\lim \limits_{n \to \infty}  ((-1)^{2n} + \dfrac{1}{n^2}) = (-1)^{2 \cdot \infty} + \dfrac{1}{\infty^2} = 1 + 0 = 1$$

Проверим подстановкой:

In [20]:
seq4 = []
for n in range(1, 10**3):
    seq4.append((-1)**(2 * n) + 1/(n**2))
seq4[0:5]

[2.0, 1.25, 1.1111111111111112, 1.0625, 1.04]

* Последовательность монотонно убывающая. 
* Снизу ограничена единицей, а сверху двойкой.
* Пятый по счету член = 1.04.

### 2. Найти 12-й член заданной неявно последовательности

$$ a_1 = 128, \ \ a_{n + 1} - a_n = 6 $$


Подставим значение и упростим последовательность: 

$$ \{a_n\}_{n=1}^{\infty} = 128 + 6 \cdot n $$

In [22]:
# Найдем 12 последовательность

seq5 = []
for n in range(1, 10**2):
    seq5.append(122 + 6 * n)
print(seq5[0:15])
print(f'12-й член последовательности равен {seq5[11]}')

[128, 134, 140, 146, 152, 158, 164, 170, 176, 182, 188, 194, 200, 206, 212]
12-й член последовательности равен 194


### 3. \* На языке Python предложить алгоритм вычисляющий численно предел $\lim \limits_{n \to +\infty} \dfrac{n}{\sqrt[n]{n!}} $ с точностью $\epsilon = 10^{-7}$


In [ ]:
%%time
seq6 = []
lim = 171
accuracy = 10**-7

for n in range(1, lim):
    seq6.append(n / math.factorial(n) ** (1 / n))  
    if n > 3:
        if seq6[n - 1] - seq6[n - 2] < accuracy:
            print(f'Значение предела с необходимой точностью получено на шаге n = {n}')
            break
seq6[lim-4:]

### 4. *Предложить оптимизацию алгоритма, полученного в задании 3, ускоряющую его сходимость.

Данный в задании предел не что иное как число e — основание натурального логарифма. Число e можно выразить через сумму ряда 
$$ \{a_n\}_{n=0}^{\infty} \dfrac{1}{n!}$$

In [ ]:
%%time
seq = []
sumseq, sumseq_last = 0, 0
lim = 100
accuracy = 10**-7

for n in range(0, lim):
    sumseq_last = sum(seq)
    seq.append(1 / math.factorial(n))  
    sumseq = sum(seq)
    if n > 3:
        
        if (sumseq - sumseq_last) < accuracy:
            print(f'Ряд сошелся с необходимой точностью на шаге n = {n}')
            break
    
sum(seq)